library/common/extensions/cert_validator/platform_bridge/platform_bridge_cert_validator.cc

```diff
@@ -0,0 +1,167 @@
+    const Network::TransportSocketOptionsConstSharedPtr& transport_socket_options,
+    SSL_CTX& /*ssl_ctx*/, const CertValidator::ExtraValidationContext& /*validation_context*/,
+    bool is_server, absl::string_view host_name) {
+  ASSERT(!is_server);
+  if (sk_X509_num(&cert_chain) == 0) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(
+          Envoy::Ssl::ClientValidationStatus::NotValidated);
+    }
+    const char* error = "verify cert chain failed: empty cert chain.";
+    stats_.fail_verify_error_.inc();
+    ENVOY_LOG(debug, error);
+    return {ValidationResults::ValidationStatus::Failed, absl::nullopt, error};
+  }
+  if (callback == nullptr) {
+    callback = ssl_extended_info->createValidateResultCallback();
+  }
+
+  std::vector<envoy_data> certs;
+  for (uint64_t i = 0; i < sk_X509_num(&cert_chain); i++) {
+    X509* cert = sk_X509_value(&cert_chain, i);
+    unsigned char* cert_in_der = nullptr;
+    int der_length = i2d_X509(cert, &cert_in_der);
+    ASSERT(der_length > 0 && cert_in_der != nullptr);
+
+    absl::string_view cert_str(reinterpret_cast<char*>(cert_in_der),
+                               static_cast<size_t>(der_length));
+    certs.push_back(Data::Utility::copyToBridgeData(cert_str));
+    OPENSSL_free(cert_in_der);
+  }
+
+  absl::string_view host;
+  if (transport_socket_options != nullptr &&
+      !transport_socket_options->verifySubjectAltNameListOverride().empty()) {
+    host = transport_socket_options->verifySubjectAltNameListOverride()[0];
+  } else {
+    host = host_name;
+  }
+  PendingValidation validation(*this, std::move(certs), host, std::move(transport_socket_options),
+                               std::move(callback));
+  auto insert_result = validations_.insert(std::move(validation));
+  ASSERT(insert_result.second);
+  PendingValidation& ref = const_cast<PendingValidation&>(*insert_result.first);
+  std::thread verification_thread(&PendingValidation::verifyCertsByPlatform, &ref);
+  std::thread::id thread_id = verification_thread.get_id();
+  validation_threads_[thread_id] = std::move(verification_thread);
+  return {ValidationResults::ValidationStatus::Pending, absl::nullopt, absl::nullopt};
+}
+
+void PlatformBridgeCertValidator::verifyCertChainByPlatform(
+    std::vector<envoy_data>& cert_chain, const std::string& host_name,
+    const std::vector<std::string>& subject_alt_names, PendingValidation& pending_validation) {
+  ASSERT(!cert_chain.empty());
+  ENVOY_LOG(trace, "Start verifyCertChainByPlatform for host {}", host_name);
+  // This is running in a stand alone thread other than the engine thread.
+  envoy_data leaf_cert_der = cert_chain[0];
+  bssl::UniquePtr<X509> leaf_cert(d2i_X509(
+      nullptr, const_cast<const unsigned char**>(&leaf_cert_der.bytes), leaf_cert_der.length));
+  envoy_cert_validation_result result =
+      platform_validator_->validate_cert(cert_chain.data(), cert_chain.size(), host_name.c_str());
```

Comment:
I tried to remove registerApi, but run into a very subtle difficulty in tests where we run cronet engine in MacOS, i.e. https://github.com/envoyproxy/envoy-mobile/actions/runs/3003635334/jobs/4822128405. If the `platform_validator_` API gets loaded based on platform, those tests will not be able to load `verify_x509_cert_chain`. And to make things more complex, one day we will have IOS validation API. Those tests will be very confused about which API to load if the engine doesn't register the right one during initialization.  I think java_tests_mac CI is the only blocker for me to get rid of registerApi stuff in this PR. Any thoughts?

Reply:
For the record, not planning to keep blocking on this - just want to ask some questions so that I understand the current situation better. Thank you :)

The logs from https://github.com/envoyproxy/envoy-mobile/actions/runs/3003635334/jobs/4822128405 are expired - I cannot see them :( .

>If the platform_validator_ API gets loaded based on platform, those tests will not be able to load verify_x509_cert_chain

Since I do not have access to the logs of the failing tests - could you elaborate on this? Any chance that you remember which tests were failing specifically? Some of our existing platform specific JNI utility methods branch on `#if(_ANDROID_API_)` check ([example here](https://github.com/envoyproxy/envoy-mobile/blob/main/library/common/jni/android_jni_utility.cc#L19-L48)) but not sure whether this would be useful in your case.

>one day we will have IOS validation API

yes, that's when the API register will come in handy. I just wonder whether for now we could avoid that additional complexity as between now and then (when we have iOS platform cert validation) a lot can change.

Reply:
> Since I do not have access to the logs of the failing tests - could you elaborate on this? Any chance that you remember which tests were failing specifically?

Some tests under https://github.com/envoyproxy/envoy-mobile/tree/main/test/java/org/chromium/net/testing runs by `android_tests / java_tests_mac` CI, and they uses cronet engines which will always use platform API for cert validation.

>  Some of our existing platform specific JNI utility methods branch on `#if(_ANDROID_API_)` check ([example here](https://github.com/envoyproxy/envoy-mobile/blob/main/library/common/jni/android_jni_utility.cc#L19-L48)) but not sure whether this would be useful in your case.

The AndroidNetworkLibrary APIs has mock interfaces, so that even on non-Android platform we can still run Java tests which call into `verify_x509_cert_chain` as long as those tests setup the mock interface.  And there is no need of `#if(_ANDROID_API_)` for all the verification related JNI calls. But if we want to call validation interfaces based on 
`#if not defined(__APPLE__)` or `#if(_ANDROID_API_)`, the cronet engine tests running on MacOS will not be able to make the right JNI calls.

That being said, today we are not supporting iOS cert validation APIs, so a temporary solution to make java tests on MacOS happy could be hard coding those JNI calls instead of calling them in `#ifdef` block. As a result, any engine enables platform based cert validation will call into those JNI functions. But this PR doesn't enable swift engine to use platform based cert validation, and in C++ engine we check platform before enabling it. This won't work once we support iOS cert validation API but still want to run tests under https://github.com/envoyproxy/envoy-mobile/tree/main/test/java/org/chromium/net/testing on MacOS.

> > one day we will have IOS validation API
> 
> yes, that's when the API register will come in handy. I just wonder whether for now we could avoid that additional complexity as between now and then (when we have iOS platform cert validation) a lot can change.

Will there be any chance not to run tests under https://github.com/envoyproxy/envoy-mobile/tree/main/test/java/org/chromium/net/testing on MacOS in the future? If that's the case, we won't even need platform registry after we support iOS cert validation API. Those CIs are helpful for Android developer using Mac, but not very valuable in real device scenario.